terraform/evaluate.go

```diff
@@ -619,13 +620,39 @@ func (d *evaluationStateData) GetResource(addr addrs.Resource, rng tfdiags.Sourc
 	rs := d.Evaluator.State.Resource(addr.Absolute(d.ModulePath))
 
 	if rs == nil {
-		// we must return DynamicVal so that both interpretations
-		// can proceed without generating errors, and we'll deal with this
-		// in a later step where more information is gathered.
-		// (In practice we should only end up here during the validate walk,
-		// since later walks should have at least partial states populated
-		// for all resources in the configuration.)
-		return cty.DynamicVal, diags
+		switch d.Operation {
+		case walkPlan:
+			// During plan as we evaluate each removed instance they are removed
+			// from the temporary working state. Since we know there there are
+			// no instances, and resources might be referenced in a context
+			// that needs to be known during plan, return an empty container of
+			// the expected type.
+			switch {
+			case config.Count != nil:
+				return cty.EmptyTupleVal, diags
+			case config.ForEach != nil:
+				return cty.EmptyObjectVal, diags
+			default:
+				// FIXME: try to prove this path should not be reached during plan.
+				//
+				// while we can reference an expanded resource with 0
+				// instances, we cannot reference instances that do not exist.
+				// Since we haven't ensured that all instances exist in all
+				// cases (this path only ever returned unknown), only log this as
+				// an error for now, and continue to return a DynamicVal
+				log.Printf("[ERROR] unknown instance %q referenced during plan", addr.Absolute(d.ModulePath))
+				return cty.DynamicVal, diags
+			}
+
+		default:
+			// we must return DynamicVal so that both interpretations
+			// can proceed without generating errors, and we'll deal with this
```

Comment:
I recognize that this is just a moved comment, but do you happen to know what "both interpretations" refers to here? I can't figure it out.

Reply:
Good question, and I don't know! I assume it's referring to the fact there there may actually be no instances, or the number of instances is unknown during validate. I added the FIXME so that we can come back and audit this a little more closely later.